tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -603,13 +603,26 @@ void SpirvEmitter::HandleTranslationUnit(ASTContext &context) {
     // Run optimization passes
     if (theCompilerInstance.getCodeGenOpts().OptimizationLevel > 0) {
       std::string messages;
-      if (!spirvToolsOptimize(&m, &messages)) {
-        emitFatalError("failed to optimize SPIR-V: %0", {}) << messages;
-        emitNote("please file a bug report on "
-                 "https://github.com/Microsoft/DirectXShaderCompiler/issues "
-                 "with source code if possible",
-                 {});
-        return;
+      // Temporarily turn off optimization when the rich debug information is
+      // generated.
+      if (spirvOptions.debugInfoRich) {
+        if (!spirvToolsLegalize(&m, &messages)) {
```

Comment:
isn't `spirvToolsLegalize` already called on line 591?

Reply:
Is it? I did not know that. Does it mean the upstream DXC will run the legalization and then the optimization when `theCompilerInstance.getCodeGenOpts().OptimizationLevel > 0` is true?

Reply:
**Regardless of `OptimizationLevel`**, if `needsLegalization` is `true`, it will run legalization. (unless legalization is disabled using `-fcgl`)

Reply:
When I test it using Sascha William examples, there are some cases it does not run legalization but it runs the optimization. I think we have to enforce it to run legalization instead of optimization when the optimization option is enabled.